scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,192 @@
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let next_window_begin = *receiver.borrow();
+
+            // We are given timestamp of the beginning of the next read window,
+            // so we have to create a checkpoint with lower timestamp.
+            let checkpoint_timestamp =
+                Duration::checked_sub(next_window_begin, Duration::from_millis(1)).ok_or(
+                    anyhow::anyhow!(
+                        "Couldn't create a checkpoint, because starting timestamp was too low ( less than 1 ms )."
+                    ),
+                )?;
+            checkpoint_maker
+                .make_checkpoint(checkpoint_timestamp)
+                .await
+                .unwrap();
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for creating checkpoints.
+#[async_trait]
+pub trait CDCCheckpointMaker: Send + Sync {
+    /// Creates new checkpoint with given timestamp.
+    async fn make_checkpoint(&mut self, timestamp: Duration) -> anyhow::Result<()>;
+}
+
+/// Default implementation for CDCCheckpointMaker trait.
+pub struct CDCDefaultCheckpointMaker {
+    session: Arc<Session>,
+    keyspace: String,
+    checkpoint_tablename: String,
+    prepared_stmt: PreparedStatement,
+}
+
+impl CDCDefaultCheckpointMaker {
+    /// Creates new CDCDefaultCheckpointMaker.
+    /// If given None value as table name,
+    /// it uses default value "cdc_checkpoint".
+    pub async fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        tablename: Option<String>,
+    ) -> anyhow::Result<Self> {
+        let session = Arc::clone(session);
+
+        const DEFAULT_TABLE: &str = "CDC_CHECKPOINT";
+        let checkpoint_tablename = tablename.unwrap_or_else(|| DEFAULT_TABLE.to_string());
+
+        CDCDefaultCheckpointMaker::create_table(&session, &keyspace, &checkpoint_tablename).await?;
+        session.await_schema_agreement().await?;
+
+        let prepared_stmt = session
+            .prepare(format!(
+                "INSERT INTO {}.{} (time) VALUES(?)",
+                &keyspace, &checkpoint_tablename
+            ))
+            .await?;
+        let cp_maker = CDCDefaultCheckpointMaker {
+            session,
+            keyspace,
+            checkpoint_tablename,
+            prepared_stmt,
+        };
+
+        Ok(cp_maker)
+    }
+
+    // Creates new table with specified schema in the given keyspace with provided name.
+    // If a table with such name already exists, doesn't do anything.
+    async fn create_table(
+        session: &Session,
+        keyspace: &str,
+        checkpoint_tablename: &str,
+    ) -> anyhow::Result<()> {
+        let schema = format!(
+            "CREATE TABLE IF NOT EXISTS {}.{} (
+            time timestamp PRIMARY KEY,
```

Comment:
I'm not sure if this is meant to be the final schema, but in its current state it's not going to work well. After a long time, you will have lots of timestamps in the table. Partition keys are not sorted by value, so on restart you will have to fetch all of them and compute the maximum manually.

Moreover, we would like to be able to save progress on a per-stream basis. Creating a separate table for each stream isn't really feasible, instead you should keep the last timestamp for each stream in a separate partition.

Reply:
Information about the current generation should also be kept, so that - on start - the log reader knows which streams to read from and which streams' last read timestamps to ask for.